ql/src/java/org/apache/hadoop/hive/ql/parse/CalcitePlanner.java

```diff
@@ -3914,33 +3914,24 @@ private RelNode genOBLogicalPlan(QB qb, Pair<RelNode, RowResolver> selPair,
       // 1. OB Expr sanity test
       // in strict mode, in the presence of order by, limit must be
       // specified
-      Integer limit = qb.getParseInfo().getDestLimit(dest);
-      if (limit == null) {
+      ASTNode limitExpr = qb.getParseInfo().getDestASTLimit(dest);
```

Comment:
Ok, so I think I have a little bit of a small design preference here that I like better.  Take it for what it's worth...

I like having related variables in one class.  We now have two variables representing limit, the limitValue and limitExpr and they both have to be checked each time.   (offset too).  

My preference would be to put this in a class like "LimitOffsetExpr" where it can hold both the expression and value.

Furthermore, we can place the  genValueFromConstantExpr method into that class to retrieve the calculated expression and removes this method from a way too cluttered up CalcitePlanner class that desperately needs a rewrite (and prolly call it something different).

Reply:
I agree with Steve's comment. Moreover, the `limitValue` is a special case of `limitExpr` so I was thinking that it should be possible to keep only the latter; same for offset.